### ISSUE TITLE
SQL current time

### ISSUE THREAD

**Issue by jakejscott** [MaterializeInc/materialize#2111]
It says in the SQL docs that the `now` function represents the time at which this dataflow was created. Is there any way to get the current time? I'm wondering if it is possible that when the current time is between a min and max date you could automatically notify subscribers via a sink. Or would you need to join a stream that emits the current time every n number of minutes?

**Comment by frankmcsherry** [comment]
Hello!

So, real talk is that this is harder than it seems. View maintenance is relatively easier when things aren't changing, and "current time" changes all the time. Often times, logic involving times ends up corresponding to windows or other ways of driving the view maintenance around. 

I'm not certain I understand your ask though: writing to a sink when the current time is in some interval seems like something you could do without materialize. Can you take another swing at an example use?

**Comment by rjnn** [comment]
Hi @jakejscott, to add some flavor, I think your idea of 'now() as a system-level source' is the right intuition for building the streaming equivalent of '`now()` as a built-in function'. A SQL function is meant to be evaluated once, making a one-off call to get the current time a useful feature, whereas a materialized view definition is meant to be continually re-evaluated.

There currently isn't a way to achieve this in Materialize without using an outside driver (i.e. pull from a source that's just printing the current time continuously), but it seems at first like a fair ask.

Thinking it through a little bit more out loud: The concept of adding a variety of 'system sources' seems like a useful one, with `current_time` being one of those sources. And your intuition of joining against it also seems like the correct way to think about how to interact with it. So if you want e.g. to send stock positions to a sink at the end of the day based on "the days closing prices", then one _could_ write a query where

`CREATE VIEW prices AS SELECT ...` (whatever is the price computation view definition)
`CREATE VIEW closing_day_prices AS SELECT * FROM prices LEFT JOIN mz_catalog.system_time WHERE extract(hour FROM mz_catalog.system_time) = 17`.
`CREATE SINK emit_closing_day_prices AS ...`

^ The above is not meant to be runnable SQL, please do not bikeshed.

The resulting behavior should be that the `emit_closing_day_prices` sink should blurt out the prices once a day at 5pm.

Now let me switch hats and devil's advocate _against_ this feature. Times are important, but it's dangerous to have Materialize decide on what the time _is_. Materialize has no opinion on the time for a few reasons: most importantly, we would like Materialize to be purely deterministic in execution, and having Materialize use the system clock is a choice that would require expensive, repeatable coordination across clusters, across replication groups (coming soon!).

TL;DR: It's much wiser to have an external clock source that informs Materialize what the time _is_. Deciding on "what is the time" is a fraught problem in distributed systems, one that we'd rather punt on.

Now if you're just working on a single system, this is quite easy to cook up yourself:

Start emitting seconds (or whatever fidelity you want) by running the bash command `while true; do date >> /tmp/date; sleep 1; done`.

Create a file source on this file with:
`CREATE SOURCE t FROM FILE '/tmp/date' WITH (tail = true) FORMAT CSV WITH 6 columns delimited by ' ';`

For the purposes of our demonstration, I also ran `echo 'hello' > /tmp/hello'` which you can read in with ` create source hello from file '/tmp/hello' with (tail = true) format text;`, or use whatever other source you have lying around. The contents of this source are unimportant, what's important is that we are going to emit the latest value of it every 10 seconds.

Now create the view:
`create view every_second as select (substring(column4, 7, 8)::int) as second from t`
(As an aside, I am not happy that I have to run `substring`, rest assured that we will add in the built-in function for simple casts to timestamp types, meanwhile I've filed #2127 to track.)

Finally, run `create materialized view emits as select text from hello left join every_second on true`

Run `tail emits`, and you should see the latest contents of 'hello' (or whatever view you substitute for hello) printed. You can then pipe this to a file SINK.

Change the view definitions for `every_second` to only emit at 5pm, or once every minute, or whatever.

In production, you'd replace that bash script with some trusted 'true' source of time (perhaps, indeed, Google's _truetime_ if it was exposed as a streaming API).

Does this make sense?

@benesch: perhaps it makes sense to add syntactic sugar to make the above steps easier? What if the syntactic sugar _wrote_ (in a persistent way) the time down in a file sink and then spun it back in as a source? Would this make it more painless? Thoughts?

**Comment by rjnn** [comment]
Here's an asciinema of it in action:
[![asciicast](https://asciinema.org/a/ZwAhsGFnR63DkGRbgp2B1YWe7.svg)](https://asciinema.org/a/ZwAhsGFnR63DkGRbgp2B1YWe7)

**Comment by rjnn** [comment]
Frank just pointed out over dinner that it can't be a left join, it has to be an inner join, as otherwise when the data side updates, we'd re-emit all the data.

You can also chain an order by limit at the end so that we aren't materializing the entire history of the data as well, and keep the memory footprint tight by just remembering the latest values. I'm sure there are a bunch more improvements to be made here, but this is the rough gist of things to get you started!

**Comment by jakejscott** [comment]
@rjnn Thanks for the detailed reply its helpful. I think streaming data into materialized at the right time would work for my use case.

> Can you take another swing at an example use?

I've being thinking about a system where a user has to trigger a bunch of events before they qualify to receive an offer, but the offer is only available to users on certain days (or window of time) and configured by a marketing team. The other use case is basically the same but the user will receive a different loyalty points multiplier which is also configured by a marketing team.

So my idea was that you would be able to express the rules the user has to perform using materialized view, but I need a way to modify this as the configured rewards days / multipliers change. So I could just do this by running something that polls the rewards config db and when the current time is between the rewards start date it streams it to materialized.

**Comment by quodlibetor** [comment]
I have some brainstorming thoughts around query-time for this in https://github.com/MaterializeInc/materialize/issues/978#issuecomment-567248845 . It would not work for pushing to sinks live, though.

That @jakejscott for the use-cases you describe I believe just embedding the event time in the data coming into materialize and joining against the rewards db tables should allow you to decide at time of ingest whether they now qualify. You don't really want to trust the time of materialized reading the event anyway.

Something sort of like:

```sql
CREATE MATERIALIZED VIEW users_with_rewards AS
SELECT user_events.user_id, sum(user_events.interaction)
FROM
    user_events
JOIN
    rewards_rules ON user_events.reward_type = rewards_rules.event_type
WHERE user_events.time >= rewards_rules.start_time
  AND user_events.time <= rewards_rules.end_time
GROUP BY user_events.user_id
```

**Comment by rjnn** [comment]
I'm going to close this issue in favor of #2439 where we have some more product definition.
